### PR TITLE
Data-annotations length fix

### LINKED ISSUE #24064: Test: System.ComponentModel.DataAnnotations.Tests.MaxLengthAttributeTests/GetValidationResult_ValueNotStringOrICollection_ThrowsInvalidCastException failed with "System.Reflection.MissingMetadataException"
Opened on behalf of @Jiayili1

The test `System.ComponentModel.DataAnnotations.Tests.MaxLengthAttributeTests/GetValidationResult_ValueNotStringOrICollection_ThrowsInvalidCastException` has failed.

Assert.Throws() Failure\r
    Expected: typeof(System.InvalidCastException)\r
    Actual:   typeof(System.Reflection.MissingMetadataException): This operation cannot be carried out because metadata for the following object was removed for performance reasons:\
    \
      EETypeRva:0x0003A200\
    \
    No further information is available. Rebuild in debug mode for better information.\
    \
    

        Stack Trace:

           at System.Reflection.Runtime.TypeInfos.RuntimeNoMetadataNamedTypeInfo.get_InternalFullNameOfAssembly() in f:\dd\ndp\fxcore\CoreRT\src\System.Private.Reflection.Core\src\System\Reflection\Runtime\TypeInfos\RuntimeBlockedTypeInfo.cs:line 229
       at System.Reflection.Runtime.BindingFlagSupport.PropertyPolicies.CoreGetDeclaredMembers($RuntimeTypeInfo type, $NameFilter optionalNameFilter, $RuntimeTypeInfo reflectedType) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.Reflection.Core\src\System\Reflection\Runtime\BindingFlagSupport\PropertyPolicies.cs:line 23
       at System.Reflection.Runtime.BindingFlagSupport.QueriedMemberList$1<System.__Canon>.Create($RuntimeTypeInfo type, String optionalNameFilter, Boolean ignoreCase) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.Reflection.Core\src\System\Reflection\Runtime\BindingFlagSupport\QueriedMemberList.cs:line 123
       at System.Reflection.Runtime.TypeInfos.RuntimeTypeInfo.TypeComponentsCache.PerNameQueryCache$1<System.__Canon>.Factory(String key) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.Reflection.Core\src\System\Reflection\Runtime\TypeInfos\RuntimeTypeInfo.TypeComponentsCache.cs:line 124
       at System.Collections.Concurrent.ConcurrentUnifier$2<System.__Canon,System.__Canon>.GetOrAdd(__Canon key) in f:\dd\ndp\fxcore\CoreRT\src\Common\src\System\Collections\Concurrent\ConcurrentUnifier.cs:line 112
       at System.Reflection.Runtime.TypeInfos.RuntimeTypeInfo.Query<System.__Canon>(String optionalName, BindingFlags bindingAttr, Func$2<__Canon,Boolean> optionalPredicate) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.Reflection.Core\src\System\Reflection\Runtime\TypeInfos\RuntimeTypeInfo.BindingFlags.cs:line 192
       at System.Reflection.Runtime.TypeInfos.RuntimeTypeInfo.GetPropertyImpl(String name, BindingFlags bindingAttr, Binder binder, Type returnType, Type[] types, ParameterModifier[] modifiers) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.Reflection.Core\src\System\Reflection\Runtime\TypeInfos\RuntimeTypeInfo.BindingFlags.cs:line 122
       at System.ComponentModel.DataAnnotations.CountPropertyHelper.TryGetCount(Object value, Int32& count) in E:\A\_work\690\s\corefx\src\System.ComponentModel.Annotations\src\System\ComponentModel\DataAnnotations\MaxLengthAttribute.cs:line 132
       at System.ComponentModel.DataAnnotations.MaxLengthAttribute.IsValid(Object value) in E:\A\_work\690\s\corefx\src\System.ComponentModel.Annotations\src\System\ComponentModel\DataAnnotations\MaxLengthAttribute.cs:line 78
       at System.ComponentModel.DataAnnotations.ValidationAttribute.IsValid(Object value, $ValidationContext validationContext) in E:\A\_work\690\s\corefx\src\System.ComponentModel.Annotations\src\System\ComponentModel\DataAnnotations\ValidationAttribute.cs:line 380
       at System.ComponentModel.DataAnnotations.ValidationAttribute.GetValidationResult(Object value, $ValidationContext validationContext) in E:\A\_work\690\s\corefx\src\System.ComponentModel.Annotations\src\System\ComponentModel\DataAnnotations\ValidationAttribute.cs:line 424
       at System.ComponentModel.DataAnnotations.Tests.MaxLengthAttributeTests.<>c.<GetValidationResult_ValueNotStringOrICollection_ThrowsInvalidCastException>b__10_0() in E:\A\_work\690\s\corefx\src\System.ComponentModel.Annotations\tests\MaxLengthAttributeTests.cs:line 124
       at System.Func$1<System.__Canon>.Invoke() in Invoke:line 16707566
    
Build : Master - 20170915.01 (UWP ILC Tests)
Failing configurations:
- Windows.10.Amd64.ClientRS3-x86
  - Release
  - Debug
- Windows.10.Amd64.ClientRS3-x64
  - Release
  - Debug

Detail: https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Filc~2F/build/20170915.01/workItem/System.ComponentModel.Annotations.Tests/analysis/xunit/System.ComponentModel.DataAnnotations.Tests.MaxLengthAttributeTests~2FGetValidationResult_ValueNotStringOrICollection_ThrowsInvalidCastException

### PR DESCRIPTION
**Pull request by weitzhandler**
Fixes #24064 
See [this](https://github.com/dotnet/corefx/issues/24064#issuecomment-330002471) comment.

### CODE REVIEW

**Comment by weitzhandler** [comment]
Looks like [all tests passed](https://mc.dot.net/#/user/weitzhandler/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/6d0266071e6cb2d78f69186067629840e669a757/workItem/System.ComponentModel.Annotations.Tests).

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MaxLengthAttribute.cs [hunk 1]
```diff
@@ -122,12 +122,21 @@ public static bool TryGetCount(object value, out int count)
                 return true;
             }
 
-            PropertyInfo property = value.GetType().GetRuntimeProperty("Count");
+            PropertyInfo property = null;
+            try
+            {
+                property = value.GetType().GetRuntimeProperty("Count");
+            }
+            catch (TypeAccessException)
```
Comment (danmosemsft) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MaxLengthAttribute.cs]: As there aren't other uses of this pattern in this library, could you please add a comment like, "// On CoreRT, this property may not be enabled for reflection."

Comment (weitzhandler) [reply src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MaxLengthAttribute.cs]: Done https://github.com/dotnet/corefx/pull/24101/commits/6170e275ba59ec3b90e6b0cacd9c07f0772c8394.

**Comment by danmosemsft** [comment]
Thanks @weitzhandler !

**Comment by weitzhandler** [comment]
@danmosemsft yet I forgot to add the enclosing parenthesis to [my comment](https://github.com/dotnet/corefx/pull/24101/commits/eb6713074c1faecafd4ffcba647b37fd360eadd2#diff-7df76ccbe7dfce696e1b8e7bebd1677eR130)...

**Comment by ajcvickers** [comment]
This fix seems wrong to me because it changes the behavior when .NET Native is being used such that a nonsensical exception will be thrown which would never be thrown when .NET Native is not being used. These kinds of errors are really hard for developers to debug.

@AtsushiKan @danmosemsft Given the limitations in .NET Native, should we avoid adding code that uses Reflection in new .NET Core code going forward?  (Btw, this is truly a case where the type is completely unknown to the framework can could be almost anything.)

If we are going to continue to use Reflection in new .NET Core code, then should it be the case that:
* The runtime gives enough details to the application so the developer can correctly figure out what to add to rd.xml to get things working?
  * If the app developer doesn't do this, then it will continue to not work
* Code that uses Reflection should provide helpful message as to what is not supported on .NET Native?
  * That is, it won't just fail in some non-obvious way like in this PR.

**Comment by weitzhandler** [comment]
@ajcvickers 
We could for rethrow the exception, maybe wrapped as an inner exception to a more familiar dedicated exception adding some details.

**Comment by danmosemsft** [comment]
@AtsushiKan should probably answer that. I haven't seen a case before where we were reflecting over a type we didn't control, but surely cases must exist.

**Comment by danmosemsft** [comment]
@weitzhandler yeah I wasn't going to reset CI just for a paren :)

**Comment by danmosemsft** [comment]
@ajcvickers apologies for merging code without you - I saw that Atsushi had signed off, and read that as your name instead.

**Comment by weitzhandler** [comment]
If I may add, consider the fact that annotation attributes is unlikely to be used with low level or unknown types.

**Comment by AtsushiKan** [comment]
>should we avoid adding code that uses Reflection in new .NET Core code going forward?

You should avoid <b>unnecessary</b> use of Reflection when an easy alternative is available (a good example being using `Type.Name` in exception messages instead of `Type.ToString()`), or using Reflection to get the members of enums which are internal to your library and thus, under your control.

Some libraries are inherently dynamic in nature and won't be a easy fit for .NET Native ever. We still let the determined developer try to use them but realistically, he'll be spending lots of time getting these exceptions from the runtime and modifying their rd.xml and ending up with a lot of the very bloat that .NET Native is supposed to eschew...

**Comment by danmosemsft** [comment]
@ajcvickers would you like me to simply revert this, or is there a modification you would like?

**Comment by ajcvickers** [comment]
@danmosemsft I would like to discuss with @divega. I suspect given the answer from @AtsushiKan that we will need to revert the entire improvement since I can't think of any way to do it without Reflection, and using Reflection will be a bad experience on .NET Native. But maybe we can come up with something reasonable.

Btw, I'm really surprised by the idea that you guys don't see Reflection on arbitrary types used more often. For example, EF uses it extensively, and EF is installed in a large percentage of Visual Studio projects.

**Comment by danmosemsft** [comment]
>  you guys don't see Reflection on arbitrary types used more often. 

I guess it was more that I hadn't seen mention of this problem before. Maybe .NET Native isn't high on their radar.

**Comment by weitzhandler** [comment]
Hi guys.
I have another idea.
As this all tunnels down to the `ICountable` issue (#23578), and `HashSet<T>` is the prominent type that is not countable in the .NET, why don't we just make `HashSet<T>` implement `ICollection<T>`?
I opened a new issue for this: #24156.

I wish we are able to make the `ICountable` interface anyway, because its lack is a design flaw. I'd be curious to see what the impact of implementing it would be, at least inside this repo.

**Comment by ajcvickers** [comment]
@weitzhandler Feel free to per peruse other ideas like you suggest, but that will take time even if they are eventually agreed upon. We can revisit this issue after those ideas pan out, but for now we need to get something in that works or back out what we have already done.

**Comment by weitzhandler** [comment]
I'd say keep the current code. We made it pretty clear for the non supporting platforms I think.

**Pull request merged**